File: SSO연동가이드/samples/SsoClient.java
package com.example.sso;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.util.Base64;

/**
 * 이노티움 SSO 중앙 집중형 인증 클라이언트 (Java 11+).
 *
 * 의존성: com.fasterxml.jackson.core:jackson-databind
 *
 * 사용:
 *   SsoClient sso = new SsoClient("https://sso.innotium.com", "sso-xxx", "secret-xxx");
 *   JsonNode user = sso.login("kim", "password");
 *   long userId = user.get("userId").asLong();   // FK로 저장
 *   sso.changePassword("kim", "old", "New1!");
 *   JsonNode u = sso.getUser(userId);
 */
public class SsoClient {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Duration TIMEOUT = Duration.ofSeconds(10);

    private final String baseUrl;
    private final String clientId;
    private final String clientSecret;
    private final HttpClient http;

    private volatile String accessToken;
    private volatile Instant expiresAt = Instant.EPOCH;

    public SsoClient(String baseUrl, String clientId, String clientSecret) {
        this.baseUrl = baseUrl.replaceAll("/+$", "");
        this.clientId = clientId; this.clientSecret = clientSecret;
        this.http = HttpClient.newBuilder().connectTimeout(TIMEOUT).build();
    }

    private synchronized String token() throws Exception {
        if (accessToken == null || Instant.now().isAfter(expiresAt)) {
            String basic = Base64.getEncoder().encodeToString(
                    (clientId + ":" + clientSecret).getBytes(StandardCharsets.UTF_8));
            HttpRequest req = HttpRequest.newBuilder(URI.create(baseUrl + "/apie/sso/oauth/client-token"))
                    .header("Authorization", "Basic " + basic)
                    .POST(HttpRequest.BodyPublishers.noBody())
                    .timeout(TIMEOUT).build();
            JsonNode body = send(req);
            accessToken = body.get("access_token").asText();
            expiresAt = Instant.now().plusSeconds(body.get("expires_in").asLong() - 60);
        }
        return accessToken;
    }

    // ───── 로그인 검증 ─────
    public JsonNode login(String loginId, String password) throws Exception {
        String payload = "{\"loginId\":" + quote(loginId) + ",\"password\":" + quote(password) + "}";
        return post("/apie/sso/auth/login", payload);
    }

    // ───── 비밀번호 변경 ─────
    public JsonNode changePassword(String loginId, String currentPassword, String newPassword) throws Exception {
        String payload = "{\"loginId\":" + quote(loginId)
                + ",\"currentPassword\":" + quote(currentPassword)
                + ",\"newPassword\":" + quote(newPassword) + "}";
        return post("/apie/sso/auth/password", payload);
    }

    // ───── 조회 ─────
    public JsonNode getUser(long userId) throws Exception {
        return get("/apie/sso/users/" + userId);
    }

    public JsonNode listUsers(String keyword, Long departmentId, String status,
                               int startIndex, int pageSize) throws Exception {
        StringBuilder url = new StringBuilder("/apie/sso/users?startIndex=")
                .append(startIndex).append("&pageSize=").append(pageSize);
        if (keyword != null) url.append("&keyword=").append(urlEncode(keyword));
        if (departmentId != null) url.append("&departmentId=").append(departmentId);
        if (status != null) url.append("&status=").append(status);
        return get(url.toString());
    }

    public JsonNode getDepartment(long departmentId) throws Exception {
        return get("/apie/sso/departments/" + departmentId);
    }

    public JsonNode listDepartments(String keyword, String status,
                                     int startIndex, int pageSize) throws Exception {
        StringBuilder url = new StringBuilder("/apie/sso/departments?startIndex=")
                .append(startIndex).append("&pageSize=").append(pageSize);
        if (keyword != null) url.append("&keyword=").append(urlEncode(keyword));
        if (status != null) url.append("&status=").append(status);
        return get(url.toString());
    }

    // ───── 내부 ─────
    private JsonNode post(String path, String jsonBody) throws Exception {
        HttpRequest req = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .header("Authorization", "Bearer " + token())
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody, StandardCharsets.UTF_8))
                .timeout(TIMEOUT).build();
        return send(req);
    }

    private JsonNode get(String path) throws Exception {
        HttpRequest req = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .header("Authorization", "Bearer " + token())
                .GET().timeout(TIMEOUT).build();
        return send(req);
    }

    private JsonNode send(HttpRequest req) throws Exception {
        HttpResponse<String> res = http.send(req, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        JsonNode body = MAPPER.readTree(res.body().isEmpty() ? "{}" : res.body());
        if (res.statusCode() >= 400) {
            String key = body.has("messageKey") ? body.get("messageKey").asText() : "UNKNOWN";
            throw new SsoApiException(res.statusCode(), key, body);
        }
        return body;
    }

    private static String quote(String s) {
        return "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
    }

    private static String urlEncode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    public static class SsoApiException extends RuntimeException {
        public final int statusCode;
        public final String messageKey;
        public final JsonNode body;
        public SsoApiException(int statusCode, String messageKey, JsonNode body) {
            super("SSO API " + statusCode + " " + messageKey);
            this.statusCode = statusCode; this.messageKey = messageKey; this.body = body;
        }
    }

    public static void main(String[] args) throws Exception {
        SsoClient sso = new SsoClient(
                System.getenv().getOrDefault("SSO_BASE_URL", "https://sso.innotium.com"),
                System.getenv("SSO_CLIENT_ID"), System.getenv("SSO_CLIENT_SECRET"));
        if (args.length < 2) { System.out.println("Usage: java SsoClient <loginId> <password>"); return; }
        JsonNode u = sso.login(args[0], args[1]);
        System.out.printf("✅ %s (userId=%d, dept=%s)%n",
                u.get("name").asText(), u.get("userId").asLong(),
                u.get("primaryDepartment").get("departmentName").asText());
    }
}

File: SSO연동가이드/samples/sso-client.js
/**
 * 이노티움 SSO 중앙 집중형 인증 클라이언트 (Node.js 18+)
 *
 * const { SsoClient } = require('./sso-client');
 * const sso = new SsoClient({ baseUrl, clientId, clientSecret });
 * const user = await sso.login('kim', 'plain');
 * await sso.changePassword('kim', 'old', 'New1!');
 * const u = await sso.getUser(user.userId);
 */

class SsoApiError extends Error {
  constructor(statusCode, messageKey, raw) {
    super(`SSO API ${statusCode} ${messageKey}`);
    this.statusCode = statusCode; this.messageKey = messageKey; this.raw = raw;
  }
}

class SsoClient {
  constructor({ baseUrl, clientId, clientSecret, timeout = 10000 }) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.clientId = clientId; this.clientSecret = clientSecret; this.timeout = timeout;
    this._token = null; this._expiresAt = 0;
  }

  async _fetchToken() {
    const basic = Buffer.from(`${this.clientId}:${this.clientSecret}`).toString('base64');
    const r = await this._req('POST', '/apie/sso/oauth/client-token', { Authorization: `Basic ${basic}` });
    this._token = r.access_token;
    this._expiresAt = Date.now() + (r.expires_in - 60) * 1000;
  }

  async _auth() {
    if (!this._token || Date.now() >= this._expiresAt) await this._fetchToken();
    return { Authorization: `Bearer ${this._token}` };
  }

  async login(loginId, password) {
    return this._req('POST', '/apie/sso/auth/login',
      { ...(await this._auth()), 'Content-Type': 'application/json' },
      JSON.stringify({ loginId, password }));
  }

  async changePassword(loginId, currentPassword, newPassword) {
    return this._req('POST', '/apie/sso/auth/password',
      { ...(await this._auth()), 'Content-Type': 'application/json' },
      JSON.stringify({ loginId, currentPassword, newPassword }));
  }

  async getUser(userId) {
    return this._req('GET', `/apie/sso/users/${userId}`, await this._auth());
  }

  async listUsers({ keyword, departmentId, status, startIndex = 0, pageSize = 50 } = {}) {
    const p = new URLSearchParams({ startIndex: String(startIndex), pageSize: String(pageSize) });
    if (keyword) p.set('keyword', keyword);
    if (departmentId) p.set('departmentId', String(departmentId));
    if (status) p.set('status', status);
    return this._req('GET', `/apie/sso/users?${p}`, await this._auth());
  }

  async getDepartment(departmentId) {
    return this._req('GET', `/apie/sso/departments/${departmentId}`, await this._auth());
  }

  async listDepartments({ keyword, status, startIndex = 0, pageSize = 500 } = {}) {
    const p = new URLSearchParams({ startIndex: String(startIndex), pageSize: String(pageSize) });
    if (keyword) p.set('keyword', keyword);
    if (status) p.set('status', status);
    return this._req('GET', `/apie/sso/departments?${p}`, await this._auth());
  }

  async _req(method, path, headers, body) {
    const c = new AbortController();
    const t = setTimeout(() => c.abort(), this.timeout);
    try {
      const res = await fetch(this.baseUrl + path, { method, headers, body, signal: c.signal });
      const text = await res.text();
      let b; try { b = text ? JSON.parse(text) : {}; } catch { b = { raw: text }; }
      if (!res.ok) throw new SsoApiError(res.status, b.messageKey || 'UNKNOWN', b);
      return b;
    } finally { clearTimeout(t); }
  }
}

module.exports = { SsoClient, SsoApiError };

// ──────── 실행 예시 ────────
if (require.main === module) {
  (async () => {
    const sso = new SsoClient({
      baseUrl: process.env.SSO_BASE_URL || 'https://sso.innotium.com',
      clientId: process.env.SSO_CLIENT_ID,
      clientSecret: process.env.SSO_CLIENT_SECRET,
    });
    const [, , loginId, password] = process.argv;
    if (!loginId) { console.log('Usage: node sso-client.js <loginId> <password>'); return; }
    try {
      const u = await sso.login(loginId, password);
      console.log(`✅ ${u.name} (userId=${u.userId}, dept=${u.primaryDepartment?.departmentName})`);
    } catch (e) {
      console.error(`❌ ${e.messageKey || e.message}`);
    }
  })();
}
